### PR TITLE
Mark strings used by 'in' operator as property names

### PR DESCRIPTION
**Pull request by irinayat-MS**
Micro optimization to create PropertyStrings for strings that are known to be property names (because they are being used as such in 'in' operator). Targeted benchmarks show x2 improvement (runtime), telemetry from crawler registered no difference but also showed that the optimization's trade off of using more memory isn't a concern (same amount of strings is registered as properties with or without the optimization). Thus, the fix has a potential of improving a subset of specific usage patterns without any known broard drawbacks, so it was decided to take it.

### CODE REVIEW

File: lib/Parser/Parse.cpp [hunk 1]
```diff
@@ -8850,36 +8853,42 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
...
                     {
-                        pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
+                        pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
                     }
                 }
+                else if (pnode1->nop == knopName && nop == knopIn)
+                {
+                    PidRefStack* pidRef = pnode1->AsParseNodeName()->pid->GetTopRef();
+                    pidRef->SetIsUsedInLdElem(true);
+                }
```
Comment (irinayat-MS) [inline lib/Parser/Parse.cpp]: This is the optimization (the rest is beautification).

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -8850,36 +8853,42 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
...
+                    Assert(pnode2->AsParseNodeClass()->pnodeConstructor);
 
-                    if (!pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor->pid)
+                    if (!pnode2->AsParseNodeClass()->pnodeConstructor->pid)
                     {
-                        pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
+                        pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
                     }
                 }
+                else if (pnode1->nop == knopName && nop == knopIn)
```
Comment (irinayat-MS) [inline lib/Parser/Parse.cpp]: This is the optimization (the rest is beautification).

Comment (irinayat-MS) [reply lib/Parser/Parse.cpp]: Looked at it with Jimmy. The current optimization for load already kicks in for patterns like:

var x = "foo";
var obj = {};
if (x in obj) console.log(obj[x]);

So the scenarios that would benefit from this fix seem to be "use a kind of tagging property on objects for which only care about its existence but not the value". Is it still worth taking the fix?

---
In reply to: [187660351](https://github.com/Microsoft/ChakraCore/pull/5135#discussion_r187660351) [](ancestors = 187660351)

Comment (sethbrenith) [reply lib/Parser/Parse.cpp]: I've also written code that looks like `if ("foo" in x) use(x.foo)`. Would this optimization help that case, where the property access following the `in` is using the dot operator?

---
In reply to: [187666992](https://github.com/Microsoft/ChakraCore/pull/5135#discussion_r187666992) [](ancestors = 187666992,187660351)

Comment (irinayat-MS) [reply lib/Parser/Parse.cpp]: if ("foo" in x) won't trigger the new optimization (literal isn't a name) and use(x.foo) is already causing foo to be a property.

**Pull request merged**